File: LeetCode/src/slidingwindow/MaxConsecutiveOnesIII.java
package slidingwindow;

public class MaxConsecutiveOnesIII {
    public int longestOnes(int[] nums, int k) {
        int result = 0, count = 0;
        int left = 0, right = 0;

        while (right < nums.length) {
            if (nums[right] == 0) {
                count++;
            }
            right++;

            while (count > k) {
                if (nums[left] == 0) {
                    count--;
                }
                left++;
            }

            result = Math.max(result, right - left);
        }

        return result;
    }

    public static void main(String[] args) {
        MaxConsecutiveOnesIII solution = new MaxConsecutiveOnesIII();
        System.out.println(solution.longestOnes(new int[]{1, 1, 0, 0, 1, 1, 1, 0, 1}, 2)); // Output: 7
        System.out.println(solution.longestOnes(new int[]{1, 0, 1, 0, 0, 0, 1, 1}, 2)); // Output: 4
    }
}

File: LeetCode_Python/slidingwindow/MaxConsecutiveOnesIII.py
class MaxConsecutiveOnesIII:
    def longestOnes(self, nums: list[int], k: int) -> int:
        n = len(nums)
        left = 0
        max_length = 0
        for right in range(n):
            if nums[right] == 0:
                k -= 1
            while k < 0:
                if nums[left] == 0:
                    k += 1
                left += 1
            max_length = max(max_length, right - left + 1)
        return max_length


# Time Complexity: O(n), where n is the length of the input list nums.
# Space Complexity: O(1), as we are using only a constant amount of space.

if __name__ == "__main__":
    solution = MaxConsecutiveOnesIII()

    print("Expected Output: 6")
    print("Actual Output:", solution.longestOnes([1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0], 2))

    print("Expected Output: 10")
    print(
        "Actual Output:",
        solution.longestOnes(
            [0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1], 3
        ),
    )
